### PR TITLE
add check_y and check_x_y to check that indexes of X and y match when encoding

### LINKED ISSUE #365: transformers that take y should check X.shape[0] == y.shape[0].
X.shape[0] should be the same as y.shape[0]. If this requisite is not met, an exception of type ValueError should be raised.

https://scikit-learn.org/stable/developers/develop.html#fitting

### LINKED ISSUE #376: Encoders that are f(X, y) can produce nan results when y has non-standard index and X becomes an np.ndarray
**Describe the bug**
In encoders where the feature encoding is a function of both the feature and the target (e.g. [feature_engine.encoding.MeanEncoder](https://feature-engine.readthedocs.io/en/1.2.x/api_doc/encoding/MeanEncoder.html)), resulting encoding can become all `nan` if:
- Feature `DataFrame` `X` is transformed upstream into an `np.ndarray` by a standard `sklearn` transformer.
**and**
- Feature `DataFrame` `X` has an index other than the standard contiguous 0-based default

The error occurs in the concatenation step within the encoder, and is caused by an index mismatch resulting from the preceding array transformation. 

I have tested all the `feature_engine.encoding` transformers and only the following three have the issue:
- [MeanEncoder](https://feature-engine.readthedocs.io/en/1.2.x/api_doc/encoding/MeanEncoder.html)
- [WoEEncoder](https://feature-engine.readthedocs.io/en/1.2.x/api_doc/encoding/WoEEncoder.html)
- [PRatioEncoder](https://feature-engine.readthedocs.io/en/1.2.x/api_doc/encoding/PRatioEncoder.html)

This was first reported by @bmreiniger [here](https://github.com/feature-engine/feature_engine/issues/368#issuecomment-1028185298), and unit tests for fix will be based on his sample code [here](https://github.com/scikit-learn-contrib/category_encoders/issues/280).

**To Reproduce**
Derived from @bmreiniger code [here](https://github.com/scikit-learn-contrib/category_encoders/issues/280). Using `feature_engine.encoding.MeanEncoder` as example:

```
    import numpy as np
    import pandas as pd
  
    from feature_engine.encoding import MeanEncoder
    
    from sklearn.impute import SimpleImputer

    # test dataframe; setup for a transfromation where
    # coded version of 'x' will be a function of target 'y'
    df: pd.DataFrame = pd.DataFrame({
        'x': ['a', 'a', 'b', 'b', 'c', 'c'],
        'y': [1, 0, 1, 0, 1, 0],
    })
    # Key - "non-standard" index that is not the usual
    # contiguous range starting a t 0
    df.index = [101, 105, 42, 76, 88, 92]

    # Set up for standard pipeline/training etc.
    X: pd.DataFrame = df[["x"]]
    y: pd.Series = df["y"]

    # Will serve as a no-op whose chief purpose is to turn the
    # X into an np.ndarray
    si = SimpleImputer(strategy="constant", fill_value="a")

    # Encoder encodes X as a function of y; this is what
    # breaks down when X becomes an array and indexes don't accidentally match in final
    # concantenation
    encoder = MeanEncoder()

    # Sequence leading to issue:
    # 1) X becomes an array
    assert type(X) == pd.DataFrame
    X_2: np.array = si.fit_transform(X)
    assert type(X_2) == np.ndarray

    # 2) Encoder encodes as function of X, y
    df_result: pd.DataFrame = encoder.fit_transform(X_2, y)
    assert type(df_result) == pd.DataFrame

    # Assertion fails: breakdown in index matches causes results to be all nan
    assert all(df_result.iloc[:, 0].notnull())
``` 

**Expected behavior**
In example above, output `DataFrame` `df_result` should have meaningful encoded values, not all `nan`.

**Desktop (please complete the following information):**
 - OS: Mac OS 11.6
 - Versions: feature-engine 1.2.0, Python 3.9.10, Pandas 1.3.4, numpy 1.22.1

### PR DESCRIPTION
**Pull request by solegalli**
closes #377 
closes #376 
closes #365 

@noahjgreen295 @bmreiniger

I am bringing the changes from #377 to this new PR at the back of former PR #408

Regarding the X and y checks, I am bringing in sklearn functionality to check y and then to check X and y (taken from their check_X_y function). I think that decreases the amount of new things we need to code.

I think I captured most of what we discussed in #377 , and I brought forward the tests and ideas from @noahjgreen295 as they were or slightly reformatted and also the suggestions from @bmreiniger.

Would you guys have a look and let me know what you think?

Let me know if you are OK to merge.

Thank you so much for raising and also for the support on this issue.

### CODE REVIEW

**Comment by noahjgreen295** [comment]
Hi @solegalli - I'm sorry I'm just having git some trouble getting this branch into my cloned repo. I did the following:
`% git pull upstream checkky`
`% git switch checkky`
I'm pretty sure that what I have now locally is the correct branch (if not, let me know).
I am now trying to put a copy into my cloned repo, to push changes to there, but when I try `git push`, I get this error:
`remote: Permission to feature-engine/feature_engine.git denied to noahjgreen295.
fatal: unable to access 'https://github.com/feature-engine/feature_engine.git/': The requested URL returned error: 403`

Am I doing this the right way? Sorry for the basic git question - thanks!

File: feature_engine/dataframe_checks.py [hunk 1]
```diff
@@ -85,6 +86,108 @@ def check_X(X: Union[np.generic, np.ndarray, pd.DataFrame]) -> pd.DataFrame:
...
+    objects, checks that their indexes match. If any is a numpy array, converts to
+    pandas object with compatible index.
+
+    This transformer ensures that we can concatenate X and y using `pandas.concat`,
+    functionality needed in the encoders.
+
+    Parameters
+    ----------
+    X: Pandas DataFrame or numpy ndarray
+    y: Pandas Series or numpy ndarray
```
Comment (bmreiniger) [inline feature_engine/dataframe_checks.py:147]: some parameters not documented

File: feature_engine/dataframe_checks.py [hunk 2]
```diff
@@ -85,6 +86,108 @@ def check_X(X: Union[np.generic, np.ndarray, pd.DataFrame]) -> pd.DataFrame:
...
+    Parameters
+    ----------
+    y : pd.Series, np.array, list
+
+    multi_output : bool, default=False
+        Whether to allow 2D y (array). If false, y will be
+        validated as a vector. y cannot have np.nan or np.inf values if
+        multi_output=True.
+
+    y_numeric : bool, default=False
```
Comment (bmreiniger) [inline feature_engine/dataframe_checks.py:106]: default in docstring disagrees with code default

File: feature_engine/dataframe_checks.py [hunk 3]
```diff
@@ -85,6 +86,108 @@ def check_X(X: Union[np.generic, np.ndarray, pd.DataFrame]) -> pd.DataFrame:
...
+    Returns
+    -------
+    y: pd.Series
+    """
+
+    if y is None:
+        raise ValueError("y cannot be None.")
+
+    elif isinstance(y, pd.Series):
+        _check_y(y, multi_output=multi_output, y_numeric=y_numeric)
```
Comment (bmreiniger) [inline feature_engine/dataframe_checks.py:120]: by not assigning, you get all the checks without the conversion, but e.g. the numeric conversion also doesn't happen...

**Review by bmreiniger — COMMENTED** [review]
I was going to suggest using sklearn's `check_X_y` and/or `check_array` in a fashion similar to your usage of `_check_y` (could reduce some duplication in your `check_X`, and uses a public sklearn method instead of a private one), but there's a lot going on there for converting to array/sparse that would be a waste.  Hrm...

Otherwise, minor comments.

**Comment by solegalli** [comment]
> Hi @solegalli - I'm sorry I'm just having git some trouble getting this branch into my cloned repo. I did the following: `% git pull upstream checkky` `% git switch checkky` I'm pretty sure that what I have now locally is the correct branch (if not, let me know). I am now trying to put a copy into my cloned repo, to push changes to there, but when I try `git push`, I get this error: `remote: Permission to feature-engine/feature_engine.git denied to noahjgreen295. fatal: unable to access 'https://github.com/feature-engine/feature_engine.git/': The requested URL returned error: 403`
> 
> Am I doing this the right way? Sorry for the basic git question - thanks!

You need to push the branch to your own repo, and make a PR from there to the main repo's checkky branch.

File: feature_engine/dataframe_checks.py (hunk 3 above)
Comment (solegalli) [reply feature_engine/dataframe_checks.py:120]: Good catch, thank you! Refactoring.

File: feature_engine/dataframe_checks.py [hunk 4]
```diff
@@ -85,6 +86,108 @@ def check_X(X: Union[np.generic, np.ndarray, pd.DataFrame]) -> pd.DataFrame:
...
+    -------
+    X: Pandas DataFrame
+    y: Pandas Series
+    """
+
+    def _check_X_y(X, y):
+        X = check_X(X)
+        y = check_y(y, multi_output=multi_output, y_numeric=y_numeric)
+        check_consistent_length(X, y)
+        return X, y
```
[the remark this answers is not in the document]
Comment (solegalli) [reply feature_engine/dataframe_checks.py:170]: Yes, makes sense. The thing is, if I ran this function first, then X and y would be a pandas df and series, so I can't really check for the different scenarios any more. That's why I took this convoluted approach instead. 

How would you approach it?

[the remark this answers is not in the document]
Comment (bmreiniger) [reply feature_engine/dataframe_checks.py:184]: Oh, of course. This is fine then. You could also maybe define "the" index in cases,  _then_ do the checks/conversions?

**Comment by solegalli** [comment]
Thank you @bmreiniger @noahjgreen295 for raising and your support on this issue. Merging.

**Pull request merged**